KAFKA-18900: Implement share.acknowledgment.mode to choose acknowledgement mode

ShivsundarR opened the pull request:
*What*

- This is in continuation to
https://github.com/apache/kafka/commit/1da30bdedf457b507ed9879a78e31c9e4c74026c,
where we will only choose the acknowledgement mode based on the config
(`share.acknowledgement.mode`) and not on the basis of how the user
designs the application.
- The default value of the config is `IMPLICIT`, so if any
empty/null/invalid value is configured, then the mode defaults to
`IMPLICIT`.
- Removed AcknowledgementModes `UNKNOWN` and `PENDING` as they are no
longer required.
- Added code to ensure if the application has any unacknowledged records
in a batch in "`explicit`" mode, then it will throw an
`IllegalStateException`. The expectation is if the mode is "explicit",
all the records received in that `poll()` would be acknowledged before
the next call to `poll()`.
- Modified the `ConsoleShareConsumer` to configure the mode to
"explicit" as it was using the explicit mode of acknowledging records.

ShivsundarR committed:
Implement explicit and implicit modes completely

ShivsundarR commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java:
```diff
@@ -1085,18 +1062,18 @@ private void ensureExplicitAcknowledgement() {
-        log.warn("Invalid value for config {}: \"{}\"", ConsumerConfig.INTERNAL_SHARE_ACKNOWLEDGEMENT_MODE_CONFIG, acknowledgementModeStr);
-        return AcknowledgementMode.UNKNOWN;
+        log.warn("Invalid value for config, {}: \"{}\", setting mode to default value : IMPLICIT", ConsumerConfig.SHARE_ACKNOWLEDGEMENT_MODE_CONFIG, acknowledgementModeStr);
+        return AcknowledgementMode.IMPLICIT;
```
For the case when the config had an invalid value, the code previously logged a `WARN` log and returned `UNKNOWN`, now that we do not have `UNKNOWN` mode anymore, I can think of 2 options here:
1. To throw an `IllegalArgumentException`, as the user should only enter "implicit" or "explicit"
2. Log a warning and just assume the default value of `IMPLICIT`.
I have kept the 2nd option now, but I do feel in case of invalid config entry, we need to throw an exception so that the user will correct the entry. For empty/null values we can still assume `IMPLICIT`, but maybe we can throw exception here?

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -695,12 +694,12 @@ public class ConsumerConfig extends AbstractConfig {
+                                .define(ConsumerConfig.SHARE_ACKNOWLEDGEMENT_MODE_CONFIG,
                                         Type.STRING,
                                         null,
                                         in(null, "implicit", "explicit"),
```
I think you should be following the style of `AutoOffsetResetStrategy` or `GroupProtocol` for this configuration. In summary:
* define an enum/class called `ShareAcknowledgementMode` in `o.a.k.clients.consumer.internals`
* validate using the names of the enum constants, uppercasing and then matching so the application can use `Explicit` or `eXplicit` or whatever
* Do not permit a null value. The default would be `ShareAcknowledgementMode.IMPLICIT.name().toLowerCase(Locale.ROOT)`

ShivsundarR replied:
Thanks for the review, yes I have now set it up similar to the `AutoOffsetResetStrategy` class by defining a separate enum.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -381,18 +381,17 @@ public class ConsumerConfig extends AbstractConfig {
-     * <code>share.acknowledgement.mode</code> is being evaluated as a new configuration to control the acknowledgement mode
-     * for share consumers. It will be removed or converted to a proper configuration before release.
-     * An alternative being considered is <code>enable.explicit.share.acknowledgement</code> as a boolean configuration.
+     * <code>share.acknowledgement.mode</code> is a config to control the acknowledgement mode
```
Just `<code>share.acknowledgement.mode</code>` is sufficient. Because this config is no longer internal, it will appear in the documentation for consumer configs for free.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -381,18 +381,17 @@ public class ConsumerConfig extends AbstractConfig {
             " If set to <code>explicit</code>, the acknowledgement mode of the consumer is explicit and it must use" +
-            " <code>org.apache.kafka.clients.consumer.ShareConsumer.acknowledge()</code> to acknowledge delivery of records.";
+            " <code>org.apache.kafka.clients.consumer.ShareConsumer.acknowledge()</code> to acknowledge delivery of records." +
+            " If unset, the acknowledgement mode of the consumer is set to 'implicit' by default";
```
Maybe the final sentence would be better as `Otherwise, the acknowledgement mode is implicit.`

AndrewJSchofield commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -588,7 +588,7 @@ public void testControlRecordsSkipped() throws Exception {
         alterShareAutoOffsetReset("group1", "earliest");
         try (Producer<byte[], byte[]> producer = createProducer();
-             ShareConsumer<byte[], byte[]> shareConsumer = createShareConsumer("group1")) {
+             ShareConsumer<byte[], byte[]> shareConsumer = createShareConsumer("group1", Map.of(ConsumerConfig.SHARE_ACKNOWLEDGEMENT_MODE_CONFIG, "explicit"))) {
```
I'd like to see constants for "explicit" and "implicit" for tests, apart from situations where invalid values are being tested intentionally.

AndrewJSchofield commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -748,28 +713,33 @@ public void testExplicitAcknowledgementCommitAsyncPartialBatch() {
+            // In EXPLICIT acknowledgement mode, we are not allowed to have unacknowledged records from a batch.
+            assertThrows(IllegalStateException.class, () -> shareConsumer1.poll(Duration.ofMillis(5000)));
+
+            // Acknowledging the 3rd record
```
Hmm. Interesting. So, can I still use the `records` collection from the last successful `poll(Duration)`? What if I reset the iterator and start running through it again?

ShivsundarR replied:
Yes we could still use it as of now. I'll write an application locally and experiment a bit to see whether it is useful to allow this.

ShivsundarR replied:
Update on this : So I experimented with a mock consumer application, overall it looks good with the current setup. When the user gets the exception, they can do what the want.

- Close the consumer, in which case the unacked records are released.

- Re-attempt to acknowledge the remaining records. We will continue to get the exception until all the records are acknowledged. After that any further consumption will continue as expected.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -115,23 +115,26 @@
  * as naturally happens when the locks time out. This limit is controlled by the broker configuration property
  * {@code group.share.record.lock.partition.limit}. By limiting the duration of the acquisition lock and automatically
  * releasing the locks, the broker ensures delivery progresses even in the presence of consumer failures.
+ *
```
nit: The style of the rest of this javadoc comment is not to have blank lines before the `<p>` and to have the `<p>` on a line by itself.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -115,23 +115,26 @@
+ *
+ * <p>If the config is set to "explicit", the consumer is using <em>explicit acknowledgement</em>. In this case:
  * <ul>
+ *     <li>The application is expected to acknowledge all the records it received in the batch before the next call to ({@link #poll(Duration)}</li>
```
Let's make it stronger. `The application must acknowledge all the records it received in the batch before`... Also, the line lengths are intentionally quite short here for readability so please split this line.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -115,23 +115,26 @@
+ * <p>If the config is set to "explicit", the consumer is using <em>explicit acknowledgement</em>. In this case:
  * <ul>
+ *     <li>The application is expected to acknowledge all the records it received in the batch before the next call to ({@link #poll(Duration)}</li>
+ *     <li> If the application has some unacknowledged records before the next call to ({@link #poll(Duration)}, then the poll()
```
nit: Extra space after `<li>`

AndrewJSchofield replied:
Actually, I would remove this sentence. We can say it more clearly with fewer words.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -140,9 +143,7 @@
- * <p>The consumer can optionally use the {@code internal.share.acknowledgement.mode} configuration property to choose
- * between implicit and explicit acknowledgement, specifying <code>"implicit"</code> or <code>"explicit"</code> as required.
- * <p>
+ *
```
I think you should have a `<p>` and no blank line.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -115,23 +115,26 @@
  * </ul>
- * If the application does not call {@link #acknowledge(ConsumerRecord, AcknowledgeType)} for any record in the batch,
- * it is using <em>implicit acknowledgement</em>. In this case:
+ * If the application sets the {@code share.acknowledgement.mode} property to "implicit" or does not configure the mode, then
```
Maybe `does not set the {@code share.acknowledgement.mode} property to "explicit", then`....

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -156,6 +157,7 @@
  * This example demonstrates implicit acknowledgement using {@link #poll(Duration)} to acknowledge the records which
  * were delivered in the previous poll. All the records delivered are implicitly marked as successfully consumed and
  * acknowledged synchronously with Kafka as the consumer fetches more records.
+ * The <code>share.acknowledgement.mode</code> property is not configured, so it is set to "implicit" by default.
```
I'd omit this line.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -181,6 +183,7 @@
  *     props.setProperty(&quot;group.id&quot;, &quot;test&quot;);
  *     props.setProperty(&quot;key.deserializer&quot;, &quot;org.apache.kafka.common.serialization.StringDeserializer&quot;);
  *     props.setProperty(&quot;value.deserializer&quot;, &quot;org.apache.kafka.common.serialization.StringDeserializer&quot;);
+ *     props.setProperty(&quot;share.acknowledgement.mode&quot;, &quot;implicit&quot;);
```
And this one.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. I've done an initial round of reviews.

Please can you add to the javadoc for the `KafkaShareConsumer.acknowledge` method to say that this method can only be used for the explicit acknowledgement mode.

ShivsundarR committed:
Addressed review comments

ShivsundarR committed:
Add doc

ShivsundarR committed:
checkstyle fix

ShivsundarR committed:
minor-fix

ShivsundarR committed:
Fix failing test

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -195,12 +194,14 @@
  *
  * <h4>Per-record acknowledgement (explicit acknowledgement)</h4>
  * This example demonstrates using different acknowledgement types depending on the outcome of processing the records.
+ * Here the share.acknowledgement.mode property is set to "explicit" so the consumer must explicitly acknowledge each record.
```
"{@code share.acknowledgement.mode}"

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -116,22 +116,23 @@
- * The consumer can choose to use implicit or explicit acknowledgement of the records it processes.
- * <p>If the application calls {@link #acknowledge(ConsumerRecord, AcknowledgeType)} for any record in the batch,
- * it is using <em>explicit acknowledgement</em>. In this case:
+ * The consumer can choose to use implicit or explicit acknowledgement of the records it processes by configuring the
```
Having read it in a browser as generated javadoc, I suggest flipping this and explaining implicit mode first, and then explicit. Since implicit is the default, that's what people will get unless they set a config.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -116,22 +116,23 @@
+ *
+ * <p>If the config is set to "explicit", the consumer is using <em>explicit acknowledgement</em>. In this case:
  * <ul>
+ *     <li>The application must acknowledge all the records it received in the batch before the next call to ({@link #poll(Duration)}</li>
```
nit: Extra "(" before `@link`.

ShivsundarR committed:
Spotless apply

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ShareConsumerConfig.java:
```diff
@@ -26,7 +26,7 @@
  * The consumer configuration behavior specific to share groups.
  */
-class ShareConsumerConfig extends ConsumerConfig {
+public class ShareConsumerConfig extends ConsumerConfig {
```
This should not be a public class. It's not part of the public interface. I know that it's currently referenced in `ShareConsumerImplTest` which is in the internals package, but I don't think that reference is necessary.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java:
```diff
@@ -456,7 +443,8 @@ private enum AcknowledgementMode {
                       final int defaultApiTimeoutMs,
-                      final String groupId) {
+                      final String groupId,
+                      final String acknowledgementMode) {
```
I'd probably give this a different variable name, such as `acknowledgementModeConfig`.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareAcknowledgementMode.java:
```diff
@@ -0,0 +1,123 @@
+    /**
+     * Returns the acknowledgement mode enum.
+     */
+    public AcknowledgementMode getAcknowledgementMode() {
```
This method is unused, so I suggest removing it. And if it remains, Kafka getters don't start with `get`.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java:
```diff
@@ -381,18 +382,16 @@ public class ConsumerConfig extends AbstractConfig {
             " If set to <code>explicit</code>, the acknowledgement mode of the consumer is explicit and it must use" +
-            " <code>org.apache.kafka.clients.consumer.ShareConsumer.acknowledge()</code> to acknowledge delivery of records.";
+            " <code>org.apache.kafka.clients.consumer.ShareConsumer.acknowledge()</code> to acknowledge delivery of records." +
+            " Otherwise, the acknowledgement mode of the consumer is set to 'implicit' by default";
```
I would remove this sentence. It's already clear when you look at the generated information without this sentence.

AndrewJSchofield reviewed (changes requested):
Thanks for the updates. It needs more effort on the state of a share consumer using explicit acknowledgement which gets an exception after polling with unacknowledged records. Apart from that, it's looking good now.

ShivsundarR committed:
Addressed review comments part II

ShivsundarR committed:
Minor change

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaShareConsumer.java:
```diff
@@ -140,9 +129,17 @@
+ *     If any records in the batch were not acknowledged, an {@link IllegalStateException} is thrown.</li>
+ *     <li>The application calls {@link #close()} which attempts to commit any pending acknowledgements and
+ *     releases any remaining acquired records.</li>
+ * </ul>
```
Let's have another try at this.
```
 * <p>
 * The consumer can choose to use implicit or explicit acknowledgement of the records it processes by configuring the
 * consumer {@code share.acknowledgement.mode} property.
 * <p>
 * If the application sets the property to "implicit" or does not set it at all, then the consumer is using
 * <em>implicit acknowledgement</em>. In this mode, the application acknowledges delivery by:
 * <ul>
 *     <li>Calling {@link #poll(Duration)} without committing, which also implicitly acknowledges all of
 *     the delivered records and commits the acknowledgements to Kafka asynchronously. In this case, no exception is
 *     thrown by a failure to commit the acknowledgements.</li>
 *     <li>Calling {@link #commitSync()} or {@link #commitAsync()} which implicitly acknowledges all of
 *     the delivered records as processed successfully and commits the acknowledgements to Kafka.</li>
 *     <li>Calling {@link #close()} which releases any acquired records without acknowledgement.</li>
 * </ul>
 * If the application sets the property to "explicit", then the consumer is using <em>explicit acknowledgment</em>.
 * The application must acknowledge all records returned from {@link #poll(Duration)} using
 * {@link #acknowledge(ConsumerRecord, AcknowledgeType)} before its next call to {@link #poll(Duration)}.
 * If the application calls {@link #poll(Duration)} without having acknowledged all records, an
 * {@link IllegalStateException} is thrown. The remaining unacknowledged records can still be acknowledged.
 * In this mode, the application acknowledges delivery by:
 * <ul>
 *     <li>Calling {@link #poll(Duration)} after it has acknowledged all records, which commits the acknowledgements
 *     to Kafka asynchronously. In this case, no exception is thrown by a failure to commit the acknowledgements.</li>
 *     <li>Calling {@link #commitSync()} or {@link #commitAsync()} which commits any pending
 *     acknowledgements to Kafka.</li>
 *     <li>Calling {@link #close()} which attempts to commit any pending acknowledgements and releases
 *     any remaining acquired records.</li>
 * </ul>
 * The consumer guarantees that the records returned in the {@code ConsumerRecords} object for a specific topic-partition
 * are in order of increasing offset. For each topic-partition, Kafka guarantees that acknowledgements for the records
 * in a batch are performed atomically. This makes error handling significantly more straightforward because there can be
 * one error code per partition.
```

ShivsundarR replied:
Thanks, this looks great. I have updated this in the PR.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumerImpl.java:
```diff
@@ -673,6 +661,10 @@ private ShareFetch<K, V> collect(Map<TopicIdPartition, NodeAcknowledgements> ack
             }
+            if (acknowledgementMode == ShareAcknowledgementMode.EXPLICIT) {
+                // We cannot leave unacknowledged records in EXPLICIT acknowledgement mode, so we throw an exception to the application.
+                throw new IllegalStateException("There are unacknowledged records from the previous fetch : " + currentFetch.records());
```
It would be much better if the entire set of records was not appended. I suggest "All records must be acknowledged in explicit acknowledgement mode".

ShivsundarR replied:
Thanks, yes I had removed the records in the last update. I have now updated the error message as well.

AndrewJSchofield reviewed (changes requested):
Just a couple of comments now.

ShivsundarR commented:
@AndrewJSchofield Thanks for the review, I have addressed all the comments and updated the PR. 

- For the exception thrown on unacknowledged records, I did some experiments with a mock application and it seems to be working fine with the current code. The user can close the consumer or attempt to acknowledge the remaining records. 

- Once all the unacknowledged records are acknowledged, the `IllegalStateException` exception wont be thrown and regular consumption will resume.

I have updated a unit test reflecting the above scenario as well.

ShivsundarR committed:
Addressed comments part III

ShivsundarR committed:
Minor update

AndrewJSchofield reviewed (approved):
Thanks for the PR. Looks good to me.
